crates/bevy_reflect/src/path.rs

```diff
@@ -104,270 +114,481 @@ impl<T: Reflect> GetPath for T {
-                Token::CloseBracket => {
-                    return Err(ReflectPathError::UnexpectedToken {
-                        index: current_index,
-                        token: "]",
-                    })
-                }
-                Token::Ident(value) => {
-                    current = read_field(current, value, current_index)?;
-                }
-            }
+        for (access, current_index) in PathParser::new(path) {
+            current = access?.read_field(current, current_index)?;
         }
-
         Ok(current)
     }
 
     fn path_mut<'r, 'p>(
         &'r mut self,
         path: &'p str,
     ) -> Result<&'r mut dyn Reflect, ReflectPathError<'p>> {
-        let mut index = 0;
         let mut current: &mut dyn Reflect = self;
-        while let Some(token) = next_token(path, &mut index) {
-            let current_index = index;
-            match token {
-                Token::Dot => {
-                    if let Some(Token::Ident(value)) = next_token(path, &mut index) {
-                        current = read_field_mut(current, value, current_index)?;
-                    } else {
-                        return Err(ReflectPathError::ExpectedIdent {
-                            index: current_index,
-                        });
-                    }
-                }
-                Token::OpenBracket => {
-                    if let Some(Token::Ident(value)) = next_token(path, &mut index) {
-                        match current.reflect_mut() {
-                            ReflectMut::List(reflect_list) => {
-                                let list_index = value.parse::<usize>()?;
-                                let list_item = reflect_list.get_mut(list_index).ok_or(
-                                    ReflectPathError::InvalidListIndex {
-                                        index: current_index,
-                                        list_index,
-                                    },
-                                )?;
-                                current = list_item;
-                            }
-                            _ => {
-                                return Err(ReflectPathError::ExpectedStruct {
-                                    index: current_index,
-                                })
-                            }
-                        }
-                    } else {
-                        return Err(ReflectPathError::ExpectedIdent {
-                            index: current_index,
-                        });
-                    }
+        for (access, current_index) in PathParser::new(path) {
+            current = access?.read_field_mut(current, current_index)?;
+        }
+        Ok(current)
+    }
+}
+
+/// A path to a field within a type. Can be used like [`GetPath`] functions to get
+/// references to the inner fields of a type.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct FieldPath(Box<[(Access, usize)]>);
```

Comment:
Might also be worth mentioning in a comment here or on `PathParser` or somewhere else what the `usize`/index means. From reading, it looks like it just means the character index in the path string, but at first I thought it meant something like path depth.